File: bindings/nodejs/examples/how_tos/outputs/unlock-conditions.ts
// Copyright 2021-2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {
    Client,
    initLogger,
    Utils,
    UnlockCondition,
    AddressUnlockCondition,
    StorageDepositReturnUnlockCondition,
    ExpirationUnlockCondition,
    TimelockUnlockCondition,
    SimpleTokenScheme,
    ImmutableAccountAddressUnlockCondition,
    AccountAddress,
} from '@iota/sdk';
require('dotenv').config({ path: '.env' });

// Run with command:
// yarn run-example ./how_tos/outputs/unlock-conditions.ts

// Build outputs with all unlock conditions
async function run() {
    initLogger();

    const client = await Client.create({});

    try {
        const ed25519Address = Utils.parseBech32Address(
            'rms1qpllaj0pyveqfkwxmnngz2c488hfdtmfrj3wfkgxtk4gtyrax0jaxzt70zy',
        );

        const accountAddress = Utils.parseBech32Address(
            'rms1pr59qm43mjtvhcajfmupqf23x29llam88yecn6pyul80rx099krmv2fnnux',
        ) as AccountAddress;

        const tokenSchema = new SimpleTokenScheme(
            BigInt(50),
            BigInt(0),
            BigInt(100),
        );

        const addressUnlockCondition: UnlockCondition =
            new AddressUnlockCondition(ed25519Address);

        // Most simple output
        const basicOutput = await client.buildBasicOutput({
            unlockConditions: [addressUnlockCondition],
        });

        // Output with storage deposit return
        const basicOutputWithStorageReturn = await client.buildBasicOutput({
            unlockConditions: [
                addressUnlockCondition,
                new StorageDepositReturnUnlockCondition(
                    ed25519Address,
                    '1000000',
                ),
            ],
        });

        // Output with timelock
        const basicOutputWithTimelock = await client.buildBasicOutput({
            unlockConditions: [
                addressUnlockCondition,
                new TimelockUnlockCondition(1),
            ],
        });

        // Output with expiration
        const basicOutputWithExpiration = await client.buildBasicOutput({
            unlockConditions: [
                addressUnlockCondition,
                new ExpirationUnlockCondition(ed25519Address, 1),
            ],
        });

        // Output with immutable account unlock condition
        const foundryOutput = await client.buildFoundryOutput({
            serialNumber: 1,
            tokenScheme: tokenSchema,
            unlockConditions: [
                new ImmutableAccountAddressUnlockCondition(accountAddress),
            ],
        });

        console.log(
            JSON.stringify(
                [
                    basicOutput,
                    basicOutputWithStorageReturn,
                    basicOutputWithTimelock,
                    basicOutputWithExpiration,
                    foundryOutput,
                ],
                null,
                2,
            ),
        );
    } catch (error) {
        console.error('Error: ', error);
    }
}

void run().then(() => process.exit());

File: bindings/python/examples/how_tos/outputs/unlock_conditions.py
import json

from dotenv import load_dotenv
from iota_sdk import (
    AddressUnlockCondition,
    Client,
    Utils,
    ExpirationUnlockCondition,
    SimpleTokenScheme,
    StorageDepositReturnUnlockCondition,
    TimelockUnlockCondition,
    ImmutableAccountAddressUnlockCondition,
)

load_dotenv()

client = Client()

ed25519_address = Utils.parse_bech32_address(
    'rms1qpllaj0pyveqfkwxmnngz2c488hfdtmfrj3wfkgxtk4gtyrax0jaxzt70zy')

account_address = Utils.parse_bech32_address(
    'rms1pr59qm43mjtvhcajfmupqf23x29llam88yecn6pyul80rx099krmv2fnnux')

address_unlock_condition = AddressUnlockCondition(ed25519_address)

token_scheme = SimpleTokenScheme(50, 0, 100)

# Most simple output
basic_output = client.build_basic_output(
    unlock_conditions=[address_unlock_condition],
)
outputs = [basic_output]

# Output with storage deposit return
basic_output = client.build_basic_output(
    unlock_conditions=[
        address_unlock_condition,
        StorageDepositReturnUnlockCondition(
            1000000,
            ed25519_address,
        ),
    ],
)
outputs.append(basic_output)

# Output with timelock
basic_output = client.build_basic_output(
    unlock_conditions=[
        address_unlock_condition,
        TimelockUnlockCondition(1),
    ],
)
outputs.append(basic_output)

# Output with expiration
basic_output = client.build_basic_output(
    unlock_conditions=[
        address_unlock_condition,
        ExpirationUnlockCondition(
            1,
            ed25519_address,
        ),
    ],
)
outputs.append(basic_output)

# Output with immutable account unlock condition
foundry_output = client.build_foundry_output(
    serial_number=1,
    token_scheme=token_scheme,
    unlock_conditions=[
        ImmutableAccountAddressUnlockCondition(
            account_address,
        ),
    ],
)
outputs.append(foundry_output)

print(json.dumps(list(map(lambda o: o.to_dict(), outputs)), indent=2))
